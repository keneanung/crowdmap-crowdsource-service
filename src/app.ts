import cors from "cors";
import express, {
  Request as ExRequest,
  Response as ExResponse,
  NextFunction,
  json,
  urlencoded,
} from "express";
import rateLimit from "express-rate-limit";
import path from "node:path";
import swaggerUi from "swagger-ui-express";
import { ValidateError } from "tsoa";
import { RegisterRoutes } from "../generated/routes";
import * as swaggerJson from "../generated/swagger.json";
import { config } from "./config/values";
import {
  AuthorizationError,
  ConflictError,
  NotFoundError,
} from "./models/api/error";

export const app = express();

app.set("trust proxy", config.trustProxy);

app.use(
  urlencoded({
    extended: true,
  }),
);
app.use(json());
app.use(
  rateLimit({
    // use a 15 minute window
    windowMs: 15 * 60 * 1000,
    // allow ten requests per second
    max: 15 * 60 * 10,
    standardHeaders: true,
  }),
);
app.use("/docs", swaggerUi.serve, (_req: ExRequest, res: ExResponse) => {
  return res.send(swaggerUi.generateHTML(swaggerJson));
});

RegisterRoutes(app);

app.use(express.static(path.join(path.dirname(__dirname), "website")));

app.use(function notFoundHandler(_req, res: ExResponse) {
  res.status(404).send({
    message: "Not Found",
  });
});

app.use(function errorHandler(
  err: unknown,
  req: ExRequest,
  res: ExResponse,
  next: NextFunction,
): ExResponse | undefined {
  if (err instanceof ValidateError) {
    console.warn(`Caught Validation Error for ${req.path}:`, err.fields);
    return res.status(422).json({
      message: "Validation Failed",
      details: err.fields,
    });
  }
  if (err instanceof AuthorizationError) {
    console.warn(`Caught Authentication Error for ${req.path}:`, err.message);
    return res.status(403).json({
      message: err.message,
    });
  }
  if (err instanceof ConflictError) {
    console.error(err);
    return res.status(409).json({
      message: err.message,
    });
  }
  if (err instanceof NotFoundError) {
    console.error(err);
    return res.status(404).json({
      message: err.message,
    });
  }
  if (err instanceof Error) {
    console.error(err);
    return res.status(500).json({
      message: "Internal Server Error",
    });
  }

  next();
  return;
});
app.use(cors());
