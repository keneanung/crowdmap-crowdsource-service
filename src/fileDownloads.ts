import * as fs from "fs";
import { Readable } from "stream";
import { config } from "./config/values";

export const downloadMapVersion = async () => {
  try {
    await downloadFile(config.versionDownloadUrl, config.versionFile);
  } catch (err) {
    throw Error("Failed to download version file", {
      cause: err,
    });
  }
};

export const downloadMapFile = async () => {
  try {
    await downloadFile(config.mapDownloadUrl, config.mapFile);
  } catch (err) {
    throw Error("Failed to download map file", {
      cause: err,
    });
  }
};

const downloadFile = async (source: string, destination: string) => {
  const res = await fetch(source);

  const fileStream = fs.createWriteStream(destination);
  if (res.body === null) {
    throw Error("No body");
  }
  const readable = Readable.fromWeb(res.body);
  readable.on("error", (err) => {
    console.error(err);
    throw Error("Failed to download file", {
      cause: err,
    });
  });
  readable.pipe(fileStream);
};
