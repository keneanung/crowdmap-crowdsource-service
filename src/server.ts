import * as fs from "fs";
import { app } from "./app";
import { config } from "./config/values";
import { downloadMapFile, downloadMapVersion } from "./fileDownloads";
import { iocContainer } from "./ioc/ioc";
import { UserService } from "./services/userService";

let mapDownloadPromise;
if (!fs.existsSync(config.mapFile)) {
  mapDownloadPromise = downloadMapFile().catch((err) => {
    console.error(err);
    process.exit(1);
  });
} else {
  mapDownloadPromise = Promise.resolve();
}

let mapVersionDownloadPromise;
if (!fs.existsSync(config.versionFile)) {
  mapVersionDownloadPromise = downloadMapVersion().catch((err) => {
    console.error(err);
    process.exit(1);
  });
} else {
  mapVersionDownloadPromise = Promise.resolve();
}

const userService = iocContainer.resolve<UserService>(UserService);
const checkAdminUser = userService.getUsers().then((users) => {
  const adminUser = users.find((user) => user.name === "admin");
  if (!adminUser) {
    const apiKey = userService.generateApiKey();
    console.log(`Generated admin API Key: ${apiKey}`);
    return userService.addUser("admin", apiKey, ["site_admin", "map_admin"]);
  }
  return Promise.resolve();
});

Promise.all([
  mapDownloadPromise,
  mapVersionDownloadPromise,
  checkAdminUser,
]).then(
  () => {
    app.listen(config.port, () => {
      console.log(
        `Crowdmap service listening at http://localhost:${config.port}`,
      );
    });
  },
  (err) => {
    console.error(err);
    process.exit(1);
  },
);
