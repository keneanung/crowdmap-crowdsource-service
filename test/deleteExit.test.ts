/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import { beforeEach, expect, test } from "@jest/globals";
import request from "supertest";
import { app } from "../src/app";
import { setupChangeServiceMock } from "./iocSetup";

beforeEach(() => {
  setupChangeServiceMock();
});

test("Should accept and return exit deletion", async () => {
  await request(app)
    .post("/change")
    .send({
      type: "delete-exit",
      roomNumber: 39478,
      direction: "east",
      reporter: "Test Reporter",
    })
    .expect(201);

  await request(app)
    .get("/change")
    .expect(200)
    .expect((res) => {
      expect(res.body).toStrictEqual([
        {
          type: "delete-exit",
          roomNumber: 39478,
          direction: "east",
          reporters: 1,
        },
      ]);
    });
});

test("Should incorporate exit deletions into the map", async () => {
  await request(app).post("/change").send({
    type: "delete-exit",
    roomNumber: 39478,
    direction: "east",
    reporter: "Test Reporter",
  });

  await request(app)
    .get("/map?format=json&timesSeen=0")
    .expect(200)
    .expect("X-Map-Version", "466.1.1")
    .expect((res) => {
      const responseText = res.text;
      const map: any = JSON.parse(responseText);
      const area: any = map.areas[230];
      const room: any = area.rooms[27];
      expect(room).toMatchSnapshot();
    });
});
