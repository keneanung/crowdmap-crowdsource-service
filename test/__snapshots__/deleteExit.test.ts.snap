// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should incorporate exit deletions into the map 1`] = `
{
  "coordinates": [
    -2,
    7,
    0,
  ],
  "environment": 10,
  "exits": [
    {
      "exitId": 20171,
      "name": "north",
    },
    {
      "exitId": 13868,
      "name": "northeast",
    },
    {
      "exitId": 28837,
      "name": "northwest",
    },
    {
      "exitId": 30231,
      "name": "west",
    },
    {
      "exitId": 21236,
      "name": "south",
    },
    {
      "exitId": 36129,
      "name": "southeast",
    },
    {
      "exitId": 26343,
      "name": "southwest",
    },
    {
      "exitId": 4882,
      "name": "down",
    },
  ],
  "id": 39478,
  "name": "Upon a floating island",
  "userData": {
    "Game Area": "the Skies",
    "indoors": "y",
  },
}
`;
